### ISSUE TITLE
Field resolver should await if Defer Fetcher fetches manually

### ISSUE THREAD

**Issue by GregYeo** [sangria-graphql/sangria#196]
**Environments,**
"org.sangria-graphql" %% "sangria" % "1.0.0-RC5",
"org.sangria-graphql" %% "sangria-json4s-jackson" % "0.3.2"
"org.mariadb.jdbc" % "mariadb-java-client" % "1.5.5",
"com.typesafe.play" % "sbt-plugin" % "2.5.10"

Development in MAC OS 10.12.2
Production in Centos 7.2

**Situation,**

Generally, field's resolver can accept future. and Fetcher's future should be also.
But When i tried to do it, The Application server gets frozen.

If we define Defer Fetcher(package sangria.execution.deferred.Fetcher), we invoke it by using deferOpt, deferSeqOpt and etc.

But if we use fetch method directly, the future not processed in resolver.
(val fetch : scala.Function2[Ctx, scala.Seq[Id], scala.concurrent.Future[scala.Seq[Res]]])

It is not always happens but mostly gets frozen since second time call.
(It works sometimes in first call)

The frozen means client's new requests are not connected anymore(client timeout) and the request requested this graphQL also is not responsed in play web framework but other background service is working like akka, slick, HikariCP and etc.

**For example,**

```
Field("subscribers", OptionType(ListType(PublishSubscriberEndpointModel.outputType)), resolve = { ctx =>
      val publishId = ctx.value._id
      val results: Future[Some[Seq[PublishSubscriberEndpointModel]]]: PublishSubscriberEndpointModel.fetcherByPublish.fetch(ctx.ctx, Seq(publishId.get))
        .map( data => Some(data.map(_._2)) )

     results
})
```

It will return  Future[Some[Seq[PublishSubscriberEndpointModel]]] but can't get the result.
But if i add Await, we can get the return value it without problem.

```
Field("subscribers", OptionType(ListType(PublishSubscriberEndpointModel.outputType)), resolve = { ctx =>
      val publishId = ctx.value._id
      val results: Future[Option[Seq[PublishSubscriberEndpointModel]]]: PublishSubscriberEndpointModel.fetcherByPublish.fetch(ctx.ctx, Seq(publishId.get))
        .map( data => Some(data.map(_._2)) )

       //Add Await
      val awaitedResult:Option[Seq[PublishSubscriberEndpointModel]] Await.result(results, 2 seconds )
       awaitedResult
})
```

I'm not sure it is bug? but the future was not handled. and fixed by Await.
The other cases returning future without fetcher is used without problem.

**Comment by OlegIlyenko** [comment]
I tried to reproduce this issue, but, unfortunately, I can't. Even when I use `Fetcher.fetch` directly in resolve function, I still get correct results. This makes me wonder, why are you using it directly? It is meant to be used only by `FetcherBasedDeferredResolver` deferred resolver. By using `Fetcher.fetch` directly you are losing all of the benefits of deferred value resolution mechanism.

`Fetcher.fetch` function is exactly the same function that you provide then you provide at `Fetcher` creation time. This means the neither `Fetcher` nor `DeferredResolver` are involved at this point. If there is a bug, then it's probably somewhere in the execution engine itself. but as i mentioned, I can't reproduce the issue.

It would be great if you could double check your fetch function and make sure that it is either successfully resolved or rejected in all possible scenatios. If you still see this issue, i would appreciate if you could try to create a minimal self-contained example that reproduces this issue.

**Comment by GregYeo** [comment]
I'm sorry to be late answer.
(I went to overseas for a moment)

Thank you for testing it.
In conclusion, It is fixed without trying to do something.

I just let the source be with Await.
And, when tested it again since i read your comment, it was just working.

I'm really sorry to bother you.

just in a case, My environment's one is changed.
-> DB Layer(mariadb, couchbase) moved to Docker-machine from local docker-engine on Mac
(because Mac docker dosen't have Docker0 bridge)

Currently, I can not re-make that environment again, because I should rollback all test data in Database. If i get face it again during developing, I will make an sample repository for it. 


I think this Issue can be closed, but i attach more comment for describing why i used Fetcher.fetch directly.

1. Context Problem.
  I needed to pass changed context to fetcher, but the ways was not existed in deferred methods. But i could use if i invoke fetch directly. Not all the case but sometimes it could be used.

2. Relation Entity Scheme Problem.
  I checked most of source to use DeferredResolver. But it seems that the resolved data should have relation key(foreign Key) for HasId.
  Currently, my project doesn't have the relation key in entity directly. So i should attach the Id into resolved data. 

I use this class to each resolved data to know what was Id for HasId
`case class RelationConnector[RelationId, RelationSource](relationId:RelationId, sourceObject: RelationSource)`

Then, i extracted the relationId in Fetch
```
Fetcher[RequestGraphQLContext, RelationConnector[Long, ItemEndpointModel], Long](
    (ctx, ids) => ItemModelHandler.retrieveByPublish(ids)
  )(HasId( _.relationId))
```

But i can not use this way in Resolver because i could not find the way mapping from resolved data 

```
 Field("item", OptionType(ItemEndpointModel.outputType), resolve = { ctx =>
      val publishProductId = ctx.value._id

//   It will return  finally RelationConnector[Long, ItemEndpointModel] in Action[Ctx, Res].
//  But i want to only return ItemEndpointModel. 
//  I can not see the way for it
//    fetcherByPublishProduct.deferOpt(publishProductId)
    

//  But if i fetch directly, i can do it like below.
      fetcherByPublishProduct
      .fetch(ctx.ctx, Seq(publishProductId))
      .map(_.map(_.sourceObject)).map(_.headOption)
    })
```

I also tried it the way using Action. 
I can not imagine how the action works exactly, but it was not works also.
```
 Field("item", OptionType(ItemEndpointModel.outputType), resolve = { ctx =>
      val publishProductId = ctx.value._id

      val leafAction = Action.deferredAction[RequestGraphQLContext, RelationConnector[ Long, ItemEndpointModel]](
        fetcherByRel.deferSeq(relationByPublish,Seq(publishProductId))
      ).map(_.map(_.sourceObject))
         .map(_.headOption)

      leafAction
}
```


 I think if the foreign key can be attached during resolving(it might be used when many to many for several tables dynamically). We do not consider of re-organizing Entity Models and can make more elastic data fetcher for relation.

This is why i used Fetch directly currently. but i will replace it by my own data fetcher for heavy load. I got inspired how to write the fetcher by sangria data fetcher API. Thanks.

I am very satisfied using Sangria GraphQL. Thanks a lot.

**Comment by OlegIlyenko** [comment]
Thanks a lot for a detailed description of the challenges that you are facing. Also glad that you found the root cause for original issue 👍 

A few ideas come to my mind when I was reading your last commet:

1. The deferred value resolution and Fetch API are completely optional components. This means that you don't have to use then in all cases where you would like to access the DB or some external data source. For example, you can make a DB query directly in the field's `resolve` function. Sangria also supports a lot of different result types, like `Future`, `Try`, etc.
2. That said, there are huge advantages to using deferred value resolution mechanism, so sometimes you probably want to use it. In more non-trivial cases (where you need to provide additional arguments to your fetcher), I think it is suitable to use `DeferredResolver` directly. It provides a lot of flexibility and you can define deferred values yourself, which means that every deferred value can have an arbitrary number of arguments. Implementing a `DeferredResolver` manually may be a bit more work compared to Fetch API, but added flexibility may be worth the effort.
3. Fetch API also provides a way to fetch the relations. For example, you can do following: `products.deferRelSeq(byCategoty, "category1")`. This will fetch all products that belong to a particular category with ID `"categoty1"`. You can find more info about it [in the docs](http://sangria-graphql.org/learn/#high-level-fetch-api) in subsection "Fetching Relations". Is looks like this can be helpful in your case as well, but I'm not 100% sure. Let me know whether it fits.

**Comment by GregYeo** [comment]
Thank you for your advise.

I left how i am thinking of your comments.

1. In my case, i decided put the short code as possible in resolver. The resolver is used as controller for handling workflow. For example, Extracting Args -> Pass args and context to auth function -> Pass args to validate Function -> Pass data fetcher -> Return fetching result. So, actually i am using them and sangria data fetcher is one of them. Thank you for letting me know it(i knew it from Sangiria Graphql Document, Thanks for good documenting)
2. I checked it in document but i could not understand in practice because i did not realize what you said from the document  I will try it. 
3. Actually, I already tried it. But i am using independent fetcher for relationship. The biggest reason is loosing Type System because it is based on MAP. 

For example, in document
```
val byParent = Relation[Category, Int]("byParent", c ⇒ Seq(c.parent))
val byProduct = Relation[Category, Int]("byProduct", c ⇒ c.products)
```
Actually when we pass it to Fetcher like `products.deferRelSeq(byCategoty, "category1")`, we knew what we will get but WE DO NOT KNOW WHAT KIND OF RELATION IS(byWhat) in types. We can know the relation only by Variable name. We loose it compile time type checking.

I'm not sure you've used to use Akka Framework. It has similar problem and they tried to give type to  Message(like sangria's Relation Object). But They are still trying to implementing in experiment(several experiment cases are already tried and closed).  I know it is very hard to design for many purpose and Map style is more elastic but i'd not recommend in production. In akka, We SHOULD WRITE TEST CODE because it is very easy too broken when we do refactoring without compile time type checking. I'd strongly recommend making test code if someone uses it.

I'm not saying or proposing designing new API. just my idea for real coding. The reason why my project use Scala is compile time type checking is helpful for less test code for many cases.

However, the API is nice idea for N+1 problem and we can overcome the problem i said by pattern matching extra case handling. It depends on how the application is using it.

Thank you again for your advice, It helps me a lot.
